tla-io/src/main/scala/at/forsyte/apalache/io/tnt/TntIR.scala

```diff
@@ -0,0 +1,346 @@
+package at.forsyte.apalache.io.tnt
+
+// This package uses upickle's macros to generate serialization functions for
+// the TNT AST and associated output data.
+//
+// See https://com-lihaoyi.github.io/upickle/#Builtins for documentation on
+// upickle semi-automatic JSON serialization docs.
+
+// The `key` pacakge allows customizing the JSON key for a class tag or attribute
+// See https://com-lihaoyi.github.io/upickle/#CustomKeys
+import upickle.implicits.key
+import scala.util.Try
+
+// We use a slightly customized deserializer
+object TntDeserializer extends upickle.AttributeTagged {
+
+  // By default, upickle uses the key `"$type"` to differentiate cases of a case
+  // class, but TNT uses the key `"kind"` to encode the variant tag. This
+  // override forms a custom deserializer that is just like the default, except
+  // the value of "kind" is used to differentiate
+  override def tagName = "kind"
+
+}
+
+import TntDeserializer.{macroRW, ReadWriter => RW}
+
+class TntIRParseError(errMsg: String) extends Exception("Input was not a valid representation of the TntIR: " + errMsg)
+
+/** The JSON output produced by tntc parse */
+case class TntcOutput(
+    status: String,
+    module: TntModule,
+    // Maps source IDs to types, see the `WithId` trait
+    types: Map[Int, TntTypeScheme])
+
+object TntcOutput {
+  implicit val rw: RW[TntcOutput] = macroRW
+
+  def read(s: ujson.Readable): Try[TntcOutput] =
+    Try(TntDeserializer.read[TntcOutput](s)).recoverWith { case err: upickle.core.AbortException =>
+      scala.util.Failure(new TntIRParseError(err.clue))
+    }
+}
+
+case class TntModule(
+    id: Int,
+    name: String,
+    defs: List[TntDef])
+object TntModule {
+  implicit val rw: RW[TntModule] = macroRW
+}
+
+/** The representation of types in the type map */
+case class TntTypeScheme(
+    @key("type") typ: TntType
+    // TODO Will we need these for anything?
+    // typeVariables: List[String],
+    // rowVariables: List[String]
+  )
+object TntTypeScheme {
+  implicit val rw: RW[TntTypeScheme] = macroRW
+}
+
+/** Source IDs, used to associate expressions with their inferred types */
+trait WithID {
+  val id: Int
+}
+
+/** The representation of TNT expressions */
+sealed trait TntEx extends WithID
+
+object TntEx {
+
+  implicit val rw: RW[TntEx] =
+    RW.merge(Name.rw, Bool.rw, TntInt.rw, Str.rw, App.rw, Lambda.rw, Let.rw)
```

Comment:
I wonder whether we should prefix the names of all the classes with `Tnt`. Otherwise, IDEs would keep auto-suggesting these classes whenever somebody is typing `Name` or `Bool`. Scala itself would not have a problem here, but I would not like to have these auxiliary classes in automatic suggestions.

Reply:
Given the possibility of namespacing (`TntEx.Bool`) I dislike the redundancy of needing the Hungarian notation on all constructors. But I didn't realize this posed a meaningful problem form auto-suggest.

I'll add the prefixes.